src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs

```diff
@@ -265,6 +265,100 @@ private async Task<FormulaValue> TryHandleSetProperty(CallNode node, EvalVisitor
+
+            if (func is UserDefinedFunction userDefinedFunc)
+            {
+                return new UserDefinedFunctionAdapter(userDefinedFunc);
+            }
+
+            if (FunctionImplementations.TryGetValue(func, out AsyncFunctionPtr ptr))
+            {
+                return new AsyncFunctionPtrAdapter(ptr);
+            }
+
+            return null;
+        }
+
+        // Adapter for AsyncFunctionPtr to common invoker interface.
+        private class AsyncFunctionPtrAdapter : IFunctionInvoker
+        {
+            private readonly AsyncFunctionPtr _ptr;
+
+            public AsyncFunctionPtrAdapter(AsyncFunctionPtr ptr)
+            {
+                _ptr = ptr;
+            }
+
+            public async Task<FormulaValue> InvokeAsync(FunctionInvokeInfo invokeInfo, CancellationToken cancellationToken)
+            {
+                var args = invokeInfo.Args.ToArray();
+                var context = invokeInfo.Context;
+                var evalVisitor = invokeInfo.Runner;
+                var irContext = invokeInfo.IRContext;
+
+                var result = await _ptr(evalVisitor, context, irContext, args).ConfigureAwait(false);
+
+                return result;
+            }
+        }
+
+        // Adapter for UDF to common invoker. 
+        // This still ensures that *invoking* a UDF has the same semantics as invoking other function calls. 
+        private class UserDefinedFunctionAdapter : IFunctionInvoker
+        {
+            private readonly UserDefinedFunction _udf;
+
+            public UserDefinedFunctionAdapter(UserDefinedFunction udf)
+            {
+                _udf = udf;
+            }
+
+            public async Task<FormulaValue> InvokeAsync(FunctionInvokeInfo invokeInfo, CancellationToken cancellationToken)
+            {
+                var args = invokeInfo.Args.ToArray();
+                var context = invokeInfo.Context;
+                var evalVisitor = invokeInfo.Runner;
+
+                var udfStack = evalVisitor._udfStack;
+
+                UDFStackFrame frame = new UDFStackFrame(_udf, args);
+                UDFStackFrame framePop = null;
+                FormulaValue result = null;
+
+                try
+                {
+                    // Push this so that we have access to args. 
+                    udfStack.Push(frame);
+
+                    // https://github.com/microsoft/Power-Fx/issues/2822
+                    // This repeats IRTranslator each time. Do once and save. 
+                    (var irnode, _) = _udf.GetIRTranslator();
+
+                    evalVisitor.CheckCancel();
+
+                    result = await irnode.Accept(evalVisitor, context).ConfigureAwait(false);
+                }
+                finally
+                {
+                    framePop = udfStack.Pop();
+                }
+
+                if (frame != framePop)
+                {
+                    throw new Exception("Something went wrong. UDF stack values didn't match.");

```

Comment:
>Exception [](http://example.com/codeflow?start=30&length=9)

Is there any better Exception we could throw here?
We could add in the message some more info

Reply:
This is just moving existing code so I didn't change the logic.  cc @adithyaselv 

Note that this condition should only happen with a bug in the interpreter (and not triggered by user code).